packages/menu/src/menu.vue

```diff
@@ -192,34 +192,30 @@ export default defineComponent({
     }) => {
       const { index, indexPath } = item
       const hasIndex = item.index !== null
-      const oldActiveIndex = activeIndex.value
-
-      if (hasIndex) {
-        activeIndex.value = item.index
-      }
-
-      ctx.emit('select', index, indexPath.value, item)
+      const emitParams = ['select', index, indexPath.value, item]
 
       if (props.mode === 'horizontal' || props.collapse) {
         openedMenus.value = []
       }
 
-      if (props.router && router && hasIndex) {
+      if (!hasIndex) {
+        return
+      }
+
+      if (props.router && router) {
         let route = item.route || item.index
-        router?.push(route)
+        const routerResult = router
+          .push(route)
           .then(navigationResult => {
-            //vue-router NavigationFailureType duplicated
-            const DUPLICATE_ROUTE = 16
-            if (navigationResult && navigationResult.type !== DUPLICATE_ROUTE) {
-              activeIndex.value = oldActiveIndex
-              throw navigationResult
+            if (!navigationResult) {
+              activeIndex.value = item.index
             }
+            return navigationResult
           })
-          .catch(error => {
-            activeIndex.value = oldActiveIndex
-            throw error
-          })
-
+        ctx.emit.apply(this, emitParams.concat(routerResult))
+      } else {
+        activeIndex.value = item.index
+        ctx.emit.apply(this, emitParams)
```

Comment:
If I remembered correctly, in `setup` method we do not have access to `this`, maybe using `this` here is a wrong idea. @HerringtonDarkholme  WDYT?

Reply:
We don't have this. `ctx.emit(...emtiParams)` is ok. `(0, ctx.emit)(...emitParams)` is safer